File: frontend/src/api/solver.ts
// 排課引擎 API:pre-flight 檢查、軟約束設定、自動排課任務與進度。

import { apiGet, apiPost, apiPut } from '@/api/client'

export interface PreflightIssue {
  level: 'error' | 'warning'
  code: string
  message: string
  subject_type: string
  subject_id: number
}
export interface PreflightReport {
  semester_id: number
  semester_label: string
  ok: boolean
  error_count: number
  warning_count: number
  issues: PreflightIssue[]
  class_count: number
  teacher_count: number
  assignment_count: number
  total_periods: number
}

export interface SoftScore {
  code: string
  name: string
  weight: number
  opportunities: number
  satisfied: number
  violations: number
  penalty: number
  rate: number
  details: string[]
}
export interface SoftReport {
  total_penalty: number
  items: SoftScore[]
}

export type JobStatus = 'queued' | 'running' | 'finished' | 'failed' | 'cancelled'

export interface SolveJob {
  job_id: string
  status: JobStatus
  semester_id: number
  source_timetable_id: number
  source_name: string
  max_seconds: number
  elapsed: number
  solutions: number
  objective: number | null
  result_timetable_id: number | null
  result_name: string | null
  error: string | null
  report: SoftReport | null
}

export interface ConstraintConfig {
  semester_id: number
  daily_subject_cap: number
  teacher_daily_max: number
  teacher_consecutive_max: number
  weights: Record<string, number>
  weight_names: Record<string, string>
}

export const preflight = (semesterId: number): Promise<PreflightReport> =>
  apiGet(`/solver/preflight?semester_id=${semesterId}`)

export const getConstraintConfig = (semesterId: number): Promise<ConstraintConfig> =>
  apiGet(`/solver/config?semester_id=${semesterId}`)

export const saveConstraintConfig = (
  semesterId: number,
  body: Omit<ConstraintConfig, 'semester_id' | 'weight_names'>,
): Promise<ConstraintConfig> => apiPut(`/solver/config?semester_id=${semesterId}`, body)

export const startAutoSchedule = (
  timetableId: number,
  maxSeconds: number,
): Promise<{ job_id: string }> =>
  apiPost(`/timetables/${timetableId}/auto-schedule`, { max_seconds: maxSeconds, seed: 0 })

export const getSolveJob = (jobId: string): Promise<SolveJob> => apiGet(`/solver/jobs/${jobId}`)
export const stopSolveJob = (jobId: string): Promise<SolveJob> =>
  apiPost(`/solver/jobs/${jobId}/stop`)
export const cancelSolveJob = (jobId: string): Promise<SolveJob> =>
  apiPost(`/solver/jobs/${jobId}/cancel`)

File: backend/app/schemas/solver.py
"""排課引擎相關 schema(pre-flight 報告、軟約束設定與達成度)。"""

from pydantic import BaseModel, Field


class PreflightIssue(BaseModel):
    level: str  # error / warning
    code: str
    message: str
    subject_type: str  # teacher / class / room / assignment / semester
    subject_id: int
    detail: dict = {}


class PreflightOut(BaseModel):
    """排課前置檢查報告。ok=False 時 errors 必然非空,自動排課應被擋下。"""

    semester_id: int
    semester_label: str
    ok: bool
    error_count: int
    warning_count: int
    issues: list[PreflightIssue] = []
    # 供 UI 顯示規模
    class_count: int
    teacher_count: int
    assignment_count: int
    total_periods: int


# ── 軟約束設定與達成度(M3-3)────────────
class ConstraintConfigIn(BaseModel):
    """權重 0 = 關閉該項軟約束。設定 UI 於 v2 才做,先以 API 調整。"""

    daily_subject_cap: int = Field(default=2, ge=1, le=8)
    teacher_daily_max: int = Field(default=6, ge=1, le=12)
    teacher_consecutive_max: int = Field(default=3, ge=1, le=12)
    weights: dict[str, int] = {}


class ConstraintConfigOut(ConstraintConfigIn):
    semester_id: int
    weight_names: dict[str, str] = {}  # S1 → 「教師偏好時段」


class SoftScoreOut(BaseModel):
    code: str
    name: str
    weight: int
    opportunities: int  # 滿分
    satisfied: int      # 得分
    violations: int
    penalty: int
    rate: float
    details: list[str] = []


class SoftReportOut(BaseModel):
    total_penalty: int
    items: list[SoftScoreOut] = []
